snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+	// its descendants have not been accepted by consensus, but this block is known.
+	// For any block ID belonging to an ancestor of the last accepted block, false is returned,
+	// as the last accepted block has been accepted by consensus.
+	processing func(id ids.ID) bool
+
+	// lastAccepted returns the last accepted block of this node.
+	lastAccepted func() ids.ID
+
+	// getSnapshot returns a snapshot of the network's nodes and their last accepted blocks,
+	// or false if it fails from some reason.
+	getSnapshot func() (snapshot, bool)
+
+	// haveWeFailedCatchingUp returns whether we have not replicated enough blocks of the given snapshot
+	haveWeFailedCatchingUp func(snapshot) bool
+}
+
+type stragglerDetector struct {
+	stragglerDetectorConfig
+
+	// continuousStragglingPeriod defines the time we have been straggling continuously.
+	continuousStragglingPeriod time.Duration
+
+	// previousStragglerCheckTime is the last time we checked whether
+	// our block height is behind the rest of the network
+	previousStragglerCheckTime time.Time
+
+	// prevSnapshot is the snapshot from a past iteration.
+	prevSnapshot snapshot
+}
+
+func newStragglerDetector(
+	getTime func() time.Time,
+	log logging.Logger,
+	minConfirmationThreshold float64,
+	lastAccepted func() (ids.ID, uint64),
+	connectedValidators func() set.Set[ids.NodeWeight],
+	connectedPercent func() float64,
+	processing func(id ids.ID) bool,
+	lastAcceptedByNodeID func(ids.NodeID) (ids.ID, bool),
+) *stragglerDetector {
+	sd := &stragglerDetector{
+		stragglerDetectorConfig: stragglerDetectorConfig{
+			lastAccepted:              dropHeight(lastAccepted),
+			processing:                processing,
+			minStragglerCheckInterval: minStragglerCheckInterval,
+			log:                       log,
+			connectedValidators:       connectedValidators,
+			connectedPercent:          connectedPercent,
+			minConfirmationThreshold:  minConfirmationThreshold,
+			lastAcceptedByNodeID:      lastAcceptedByNodeID,
+			getTime:                   getTime,
+		},
+	}
+
+	sd.getSnapshot = sd.getNetworkSnapshot
+	sd.haveWeFailedCatchingUp = sd.failedCatchingUp
+
+	return sd
+}
+
+// CheckIfWeAreStragglingBehind returns for how long our ledger is behind the rest
+// of the nodes in the network. If we are not behind, zero is returned.
+func (sd *stragglerDetector) CheckIfWeAreStragglingBehind() time.Duration {
+	now := sd.getTime()
+	if sd.previousStragglerCheckTime.IsZero() {
+		sd.previousStragglerCheckTime = now
+		return 0
+	}
+
+	// Don't check too often, only once in every minStragglerCheckInterval
+	if sd.previousStragglerCheckTime.Add(sd.minStragglerCheckInterval).After(now) {
+		return 0
+	}
+
+	defer func() {
+		sd.previousStragglerCheckTime = now
+	}()
+
+	if sd.prevSnapshot.isEmpty() {
+		snapshot, ok := sd.getSnapshot()
+		if !ok {
+			sd.log.Trace("No node snapshot obtained")
+			sd.continuousStragglingPeriod = 0
+		}
+		sd.prevSnapshot = snapshot
```

Comment:
this seems not safe to set here if !ok. should it be just an empty snapshot?

Reply:
Exactly, if ! ok then it should be empty. But good point, I'll update it explicitly to be empty.